job-server/src/main/scala/spark/jobserver/io/FileCacher.scala

```diff
@@ -29,19 +29,34 @@ trait FileCacher {
 
   // Cache the jar file into local file system.
   protected def cacheBinary(appName: String,
-                          binaryType: BinaryType,
-                          uploadTime: DateTime,
-                          binBytes: Array[Byte]) {
-    val outFile =
-      new File(rootDir, createBinaryName(appName, binaryType, uploadTime))
-    val bos = new BufferedOutputStream(new FileOutputStream(outFile))
+                            binaryType: BinaryType,
+                            uploadTime: DateTime,
+                            binBytes: Array[Byte]) {
+    val targetFullBinaryName = createBinaryName(appName, binaryType, uploadTime)
+    val tempSuffix = ".tmp"
+    val tempOutFile = File.createTempFile(targetFullBinaryName + "-", tempSuffix, new File(rootDir))
+    val tempOutFileName = tempOutFile.getName
+    val bos = new BufferedOutputStream(new FileOutputStream(tempOutFile))
+
     try {
-      logger.debug("Writing {} bytes to file {}", binBytes.length, outFile.getPath)
+      logger.debug("Writing {} bytes to a temporary file {}", binBytes.length, tempOutFile.getPath)
       bos.write(binBytes)
       bos.flush()
     } finally {
       bos.close()
     }
+
+    logger.debug("Renaming the temporary file {} to the target full binary name {}",
+      tempOutFileName, targetFullBinaryName: Any)
+
+    val tempFile = new File(rootDir, tempOutFileName)
+    if( ! tempFile.renameTo(new File(rootDir, targetFullBinaryName))) {
```

Comment:
I would prefer the style `if (!tempFile.renameTo(....`
This is how most of the rest of the code is.